### PR TITLE
Add the stream upload which starts uploading parts before Flush

### PR DESCRIPTION
**Pull request by ggtakec**
### Relevant Issue (if applicable)
n/a

### Overview
In multi-part upload (mix, non-mix upload), the function to upload the file part sequentially before the file is flushed has been added.

### Details
The current s3fs will only start uploading a file when a flush is called for the file.  
This PR code has added an option called `streamupload` to allow s3fs to upload the file part before the file is flushed.  
The `streamupload` option is only effective when multipart upload (mixupload and nomixupload) is enabled.  

The individual explanations are as follows:  

#### (1) `streamupload` option
This is an option to enable the Stream upload function.  
This option is a tentative option.  
I will remove this option once this PR has been merged and fully tested.  
This function should be as the default behavior of s3fs, I plan to enable this feature like multipart upload.  
At that time, I will add the `nostream` option(pseudonym) instead, it is similar to nomultipart etc.  

#### (2) Multipart size
When Stream upload is enabled, each part size for multipart upload is fixed(specified by the `multipart_size` option).  
In other words, from the beginning of the file, the size indicated by the `multipart_size` option is used as the boundary, and each part is uploaded.  

#### (3) Part upload conditions
When all the data for the fixed range part shown in (2) is written, the upload of that part will start.(Multipart upload will start even if it is not flushed)  
If writing occurs again for the range of the part that has already been uploaded, the range will be uploaded again.  

If the written area does not fill the range of the part, the part will not be uploaded until flush is called.  
This range will be uploaded when flush is called.  

#### (4) Thread pool
The code for this additional feature is implemented to have a thread pool.  
This thread pool is used in each part's upload call.  
The thread pool is initialized when s3fs starts, and all threads are started and put into a standby state.  
Thus the `max_thread_count` option(provisional) has been added for specifying this thread pool count.  

This option is a temporary option like `streamupload`.  
This option will be replaced with the `parallel_count` option, etc., when the s3fs refurbishment(including this PR) is completed.  

#### (5) About test
Existing tests are sufficient for uploading files.  
Testing for opening files, writing to non-contiguous areas, and closing files can be done with the recently added `write_multiblock` test.  

Testing of large files was done individually, please see (6).  

#### (6) Performance
Performance comparisons involving large files are performed individually and summarized in the Gist below:  
https://gist.github.com/ggtakec/0482aca53643681e2e410ed4032b780f

The speed of uploading 5GB files has been improved by about **40%**.  

### NOTE
This PR is intended for performance tuning and source code cleanup.
The refurbishment will be a series of modifications including this PR.
In a series of fixes, I plan to use the thread pool mentioned above, and to fix downloads, HEAD requests, and so on.
And when the series of refurbishments is complete, the two tentative options mentioned above will also be sorted out.

### CODE REVIEW

File: src/fdcache_fdinfo.cpp [hunk 1]
```diff
@@ -225,25 +356,695 @@ bool PseudoFdInfo::ClearUntreated(off_t start, off_t size)
...
+                // This current area is also the uploaded area
+                //
+                // [NOTE]
+                // The uploaded area is aligned with boundary, there are all data in
+                // this current area locally(which includes all data of untreated area).
+                // So this current area only needs to be uploaded again.
+                //
+                S3FS_PRN_DBG("Cancel upload: start=%lld, size=%lld", static_cast<long long int>(overlap_uploaded_iter->startpos), static_cast<long long int>(overlap_uploaded_iter->size));
+                cancel_upload_list.push_back(*overlap_uploaded_iter);               // add this uploaded area to cancel_upload_list
+                upload_list.erase(overlap_uploaded_iter);                           // remove it from upload_list
```
Comment (gaul) [inline src/fdcache_fdinfo.cpp:957]: Does this invalidate `upload_iter`?

**Review by gaul — CHANGES_REQUESTED** [review]
First, sorry for my extended delay.  I will be more prompt for this PR from now.  I made some smaller comments but my two big concerns:

* Is the thread pool necessary?  This adds complexity that I doubt we need.
* Can we test or shrink this PR somehow?  It is hard for me to keep the entire logic in my head.

I will do some local testing as well but wanted to give some feedback today.

**Comment by ggtakec** [comment]
Thank you for your review, and I will reply to you individually, so please be patient.

First, I reply the following questions.

> Is the thread pool necessary? This adds complexity that I doubt we need.

Currently, this PR code(additional option mode) and old codes live together.
As a result, there are two request types in this PR code, the code that dynamically creates and destroys threads, and thread pool.
I think this is adding complexity, as you pointed out.

I want to that all/most processing of creating and destroying threads in the server communication(curl) part of s3fs to migrate to this thread pool.
I plan this is the goal of a series of future code fixes(about requests), including this PR.
If these can be implemented, all server requests will be queued and processed by free threads.
And I think that it can be finally put together in one logic, and it can be totally and simplified.
(At the time of this PR, it looks complicated because it has new and old)

And multipart uploads for large file and listing directories with many files create the many requests.
Then, I think that the overhead of creating and destroying threads for many requests by this one processing cannot be ignored.
Therefore, I think it is better to use a thread pool.

> Can we test or shrink this PR somehow? It is hard for me to keep the entire logic in my head.

I also wanted to divide this PR, but it is a series of operations and it is very difficult.
The `UntreatedParts` class is a pre-built class for this PR.
I tried to divide it in advance as much as possible like this, but it is difficult to divide only this PR.

And the test for this PR is run with the following test option:
`"use_cache=${CACHE_DIR} -o ensure_diskfree=${ENSURE_DISKFREE_SIZE} -o fake_diskfree=${FAKE_FREE_DISK_SIZE} -o streamupload"` (in small-integration-test.sh）

I think the full test is the same as the current master test, assuming the `streamupload` option becomes as default.
(It may be better to add boundary/limit tests such as parameters, but I think that is the same as it is now.)

And the tests that depend on the logic changes(internally) for this PR have already been added as `write_multiblock.cc`.
(This test was added to test with this PR. Of course, it is a test that is also required in the current code)

File: src/fdcache_fdinfo.cpp [hunk 2]
```diff
@@ -225,25 +356,695 @@ bool PseudoFdInfo::ClearUntreated(off_t start, off_t size)
...
+                    //
+                    // Something wrong in this unloaded area.
+                    //
+                    // This area is not aligned with the boundary, then this condition
+                    // is unrecoverable and return failure.
+                    //
+                    S3FS_PRN_ERR("The uploaded list may not be the boundary for the maximum multipart upload size. No further processing is possible.");
+                    return false;
+                }
+                // Set this iterator to ovrelap iter
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_fdinfo.cpp]: Thanks, fixed this.

File: src/threadpoolman.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+};
+
+typedef std::list<thpoolman_param*>  thpoolman_params_t;
+
+typedef std::list<pthread_t>  thread_list_t;
+
+//------------------------------------------------
+// Class ThreadPoolMan
+//------------------------------------------------
+class ThreadPoolMan
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/threadpoolman.h:58]: I believe that multiple calls will increase the overhead.
See the comments below for more information.
I marked Resolved for this review, but if there is a problem, please point it out again to the corrected code or post comments.

File: test/small-integration-test.sh [hunk 4]
```diff
@@ -52,6 +52,7 @@ if [ -n "${ALL_TESTS}" ]; then
         sigv4
         singlepart_copy_limit=10  # limit size to exercise multipart code paths
         #use_sse  # TODO: S3Proxy does not support SSE
+        "use_cache=${CACHE_DIR} -o ensure_diskfree=${ENSURE_DISKFREE_SIZE} -o fake_diskfree=${FAKE_FREE_DISK_SIZE} -o streamupload"
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply test/small-integration-test.sh:55]: See the comments below for more details on this point as well.
Since there are no external specification changes, I think we can check the operation without adding tests.
However, it may be possible to add inspections such as testing the boundaries of the upload part.
If you want to add it, it is also necessary for the existing code, so let's deal with it with another issue or PR.

[the remark this answers is not in the document]
Comment (ggtakec) [reply test/small-integration-test.sh:55]: I marked Resolved for this review, but if there is a problem, please point it out again to the corrected code or post comments.

File: src/fdcache_entity.cpp [hunk 5]
```diff
@@ -1671,6 +1720,187 @@ int FdEntity::RowFlushMixMultipart(PseudoFdInfo* pseudo_obj, const char* tpath)
...
+        filepart_list_t cancel_uploaded_list;
+        if(!pseudo_obj->ExtractUploadPartsFromAllArea(to_upload_list, to_copy_list, to_download_list, cancel_uploaded_list, S3fsCurl::GetMultipartSize(), pagelist.Size(), FdEntity::mixmultipart)){
+            S3FS_PRN_ERR("Failed to extract various upload parts list from all area: errno(EIO)");
+            return -EIO;
+        }
+
+        //
+        // Check total size for downloading and Download
+        //
+        total_mp_part_list mptoal;
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_entity.cpp]: Thanks, I changed total_mp_part_list to inline function.

File: src/fdcache_untreated.cpp [hunk 6]
```diff
@@ -274,6 +191,85 @@ bool UntreatedParts::ClearParts(off_t start, off_t size)
...
+{
+    AutoLock auto_lock(&untreated_list_lock);
+
+    for(untreated_list_t::iterator iter = untreated_list.begin(); iter != untreated_list.end(); ++iter){
+        if(iter->untreated_tag == last_tag){
+            if(0 < size){
+                iter->start = start;
+                iter->size  = size;
+            }else{
+                untreated_list.erase(iter);
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_untreated.cpp]: Thanks, I fixed.

File: src/fdcache_untreated.cpp [hunk 7]
```diff
@@ -274,6 +191,85 @@ bool UntreatedParts::ClearParts(off_t start, off_t size)
     return true;
 }
 
+//
+// Update the last updated Untreated part
+//
+bool UntreatedParts::GetLastUpdatePart(off_t& start, off_t& size)
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_untreated.cpp:197]: @gaul Sorry, please tell me the meaning of this point in a little more detail.

File: src/fdcache_fdinfo.h [hunk 8]
```diff
@@ -55,7 +94,7 @@ class PseudoFdInfo
 
         bool Set(int fd, int open_flags);
         bool ClearUploadInfo(bool is_clear_part = false, bool lock_already_held = false);
-        bool InitialUploadInfo(const std::string& id);
+        bool InitialUploadInfo(const std::string& id, bool lock_already_held = false);
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_fdinfo.h]: It have been changed it to AutoLock::Type.

File: src/fdcache_fdinfo.h [hunk 9]
```diff
@@ -22,26 +22,65 @@
...
+        volatile int            instruct_count;     // number of instructions for processing by threads
+        volatile int            completed_count;    // number of completed processes by thread
+        int                     last_result;        // the result of thread processing
 
     private:
+        static void* MultipartUploadThreadWorker(void* arg);
+
         bool Clear();
+        void CloseUploadFd(bool lock_already_held = false);
+        bool OpenUploadFd(bool lock_already_held = false);
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_fdinfo.h]: It have been changed it to AutoLock::Type.

File: src/fdcache_fdinfo.cpp [hunk 10]
```diff
@@ -20,18 +20,108 @@
...
+        return (void*)(intptr_t)(-EIO);
+    }
+    S3FS_PRN_INFO3("Upload Part Thread [tpath=%s][start=%lld][size=%lld][part=%d]", pthparam->path.c_str(), static_cast<long long>(pthparam->start), static_cast<long long>(pthparam->size), pthparam->part_num);
+
+    if(0 != pthparam->ppseudofdinfo->last_result){
+        S3FS_PRN_DBG("Already occurred error, thus this thread worker is exiting.");
+
+        AutoLock auto_lock(&(pthparam->ppseudofdinfo->upload_list_lock));
+
+        if(0 < pthparam->ppseudofdinfo->instruct_count){
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_fdinfo.cpp:62]: Yes, when this underflows, there is a fatal bug that deserves Abort.
However, I don't think that situation will occur(uh...maybe), but would you like to add processing equivalent to abort?

File: src/fdcache_fdinfo.cpp [hunk 11]
```diff
@@ -225,25 +356,695 @@ bool PseudoFdInfo::ClearUntreated(off_t start, off_t size)
...
+    if(!IsUploading()){
+        S3FS_PRN_ERR("Multipart Upload has not started yet.");
+        return false;
+    }
+    if(start < 0 || size <= 0 || part_num < 0 || !ppetag){
+        S3FS_PRN_ERR("Parameters are wrong.");
+        return false;
+    }
+
+    AutoLock auto_lock(&upload_list_lock, lock_already_held ? AutoLock::ALREADY_LOCKED : AutoLock::NONE);
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_fdinfo.cpp]: Now this method is private and when calling this method the upload_list_lock is already locked then.
However, since this method call may be added in the future, I made a lock processing to operate upload_list.

File: src/fdcache_fdinfo.cpp (hunk 1 above)
Comment (ggtakec) [reply src/fdcache_fdinfo.cpp]: Thanks, it was a bug, and I fixed it.

**Comment by ggtakec** [comment]
After rebase to master, I can't pass the test.
Maybe it's affected by the #1887 fix.
I'll try to fix it right away.

**Comment by ggtakec** [comment]
Test failed about `truncate`, so that I rebased it to the master containing #1919. Then, the error went away.

File: src/fdcache_fdinfo.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/fdcache_fdinfo.cpp:62]: It does not abort, but added error handling when underflow occurs.

**Review by gaul — APPROVED** [review]
I tested this locally and can confirm that it improves performance when the input is slow, e.g., streaming from `xz` or similar.  Let's keep this default off for at least one release to test it more.

**Comment by gaul** [comment]
Actually I see `test_truncate_upload` failures with `-o streamupload` when running rebased with master...

**Comment by ggtakec** [comment]
@gaul Thanks merging.
It's `test_external_creation`, not `test_truncate_upload`, right? (I think it will be solved by the amendment currently being applied for in PR1988.)

**Comment by gaul** [comment]
These tests both fail:

FAIL: test_truncate_upload
FAIL: test_upload_sparsefile

**Comment by ggtakec** [comment]
Thanks, I check those tests and try to fix.
